Mathlib/Computability/ContextFreeGrammar.lean

```diff
@@ -143,3 +146,77 @@ lemma Derives.eq_or_tail {u w : List (Symbol T g.NT)} (huw : g.Derives u w) :
   (Relation.ReflTransGen.cases_tail huw).casesOn (Or.inl ∘ Eq.symm) Or.inr
 
 end ContextFreeGrammar
+
+/-- Context-free languages are defined by context-free grammars. -/
+def Language.IsContextFree (L : Language T) : Prop :=
+  ∃ g : ContextFreeGrammar.{uT} T, g.language = L
+
+/-- `L.reverse` is a language that contains exactly all words from `L` backwards. -/
+def Language.reverse (L : Language T) : Language T := { w : List T | w.reverse ∈ L }
```

Comment:
should go in Mathlib.Computability.Language

Reply:
and we should add the basic lemmas for it, i.e. the reverse of the zero and 1 language, reverse of union of two languages, concatenation etc

Reply:
Please check the API.